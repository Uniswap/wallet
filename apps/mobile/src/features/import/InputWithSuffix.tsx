import { ResponsiveValue } from '@shopify/restyle'
import { useCallback, useState } from 'react'
import {
  LayoutRectangle,
  NativeSyntheticEvent,
  TextInput as NativeTextInput,
  TextInputContentSizeChangeEventData,
} from 'react-native'
import { TextInput } from 'src/components/input/TextInput'
import { Box } from 'src/components/layout'
import { IS_ANDROID } from 'src/constants/globals'
import { Theme } from 'ui/src/theme/restyle/theme'

interface Props {
  autoCorrect: boolean
  blurOnSubmit: boolean
  theme: Theme
  inputAlignment: ResponsiveValue<'center' | 'flex-start', Theme>
  value?: string
  inputFontSize: number
  inputMaxFontSizeMultiplier: number
  inputSuffix?: string
  textAlign?: 'left' | 'right' | 'center'
  textInputRef: React.RefObject<NativeTextInput>
  layout?: LayoutRectangle | null
  onBlur?: () => void
  onFocus?: () => void
  onChangeText?: (text: string) => void
  onSubmitEditing?: () => void
}

export default function InputWithSuffix(props: Props): JSX.Element {
  return IS_ANDROID && props.inputSuffix ? (
    <Box width="100%">
      <Inputs {...props} layerType="foreground" />
      <Inputs {...props} layerType="background" />
    </Box>
  ) : (
    <Inputs {...props} />
  )
}

function Inputs({
  value,
  layout,
  theme,
  inputSuffix,
  inputAlignment,
  inputFontSize,
  inputMaxFontSizeMultiplier,
  textAlign,
  textInputRef,
  layerType,
  ...inputProps
}: Props & { layerType?: 'foreground' | 'background' }): JSX.Element {
  const [isMultiline, setIsMultiline] = useState(false)

  const handleContentSizeChange = useCallback(
    (e: NativeSyntheticEvent<TextInputContentSizeChangeEventData>) => {
      if (textInputRef.current) {
        setIsMultiline(Math.floor(e.nativeEvent.contentSize.height / inputFontSize) > 1)
      }
    },
    [textInputRef, inputFontSize]
  )

  return (
    <Box
      alignItems="flex-end"
      flexDirection="row"
      justifyContent={inputAlignment}
      width="100%"
      {...(layerType === 'foreground' ? { position: 'absolute', left: 0, bottom: 0 } : {})}
      {...(layerType === 'background' ? { pointerEvents: 'none' } : {})}>
      {layerType === 'foreground' ? (
        <TextInput
          ref={textInputRef}
          backgroundColor="none"
          color="textPrimary"
          fontSize={inputFontSize}
          justifyContent="flex-start"
          lineHeight={inputFontSize}
          maxFontSizeMultiplier={inputMaxFontSizeMultiplier}
          multiline={true}
          px="none"
          py="none"
          scrollEnabled={false}
          textAlign={
            textAlign ?? (isMultiline || inputAlignment === 'flex-start' ? 'left' : 'center')
          }
          textAlignVertical="bottom"
          value={value}
          width="100%"
          onContentSizeChange={handleContentSizeChange}
          {...inputProps}
        />
      ) : (
        <TextInput
          autoFocus
          autoCapitalize="none"
          backgroundColor="none"
          color="textPrimary"
          fontSize={inputFontSize}
          justifyContent="center"
          lineHeight={inputFontSize}
          maxFontSizeMultiplier={inputMaxFontSizeMultiplier}
          multiline={true}
          px="none"
          py="none"
          returnKeyType="done"
          scrollEnabled={false}
          selectionColor={theme.colors.textPrimary}
          spellCheck={false}
          testID="import_account_form/input"
          textAlign={textAlign ?? (inputAlignment === 'center' || !value ? 'left' : 'center')}
          textAlignVertical="bottom"
          value={value}
          width={value ? 'auto' : (layout?.width || 0) + theme.spacing.spacing8}
          {...(layerType === 'background'
            ? { opacity: 0, editable: false }
            : { ...inputProps, ref: textInputRef })}
        />
      )}
      {inputSuffix && value && !value.includes(inputSuffix) ? (
        <TextInput
          backgroundColor="none"
          color="textSecondary"
          editable={false}
          fontSize={inputFontSize}
          justifyContent="center"
          lineHeight={inputFontSize}
          maxFontSizeMultiplier={inputMaxFontSizeMultiplier}
          multiline={true}
          px="none"
          py="none"
          scrollEnabled={false}
          textAlignVertical="bottom"
          value={inputSuffix}
          {...(layerType === 'foreground' ? { opacity: 0 } : {})}
        />
      ) : null}
    </Box>
  )
}
