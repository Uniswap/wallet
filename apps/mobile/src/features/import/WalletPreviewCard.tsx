import React from 'react'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { SelectionCircle } from 'src/components/input/SelectionCircle'
import { Box, Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { Unicon } from 'src/components/unicons/Unicon'
import { ChainId } from 'src/constants/chains'
import { useIsDarkMode } from 'src/features/appearance/hooks'
import { useENS } from 'src/features/ens/useENS'
import { ElementName } from 'src/features/telemetry/constants'
import { shortenAddress } from 'src/utils/addresses'
import { formatUSDPrice, NumberType } from 'src/utils/format'

interface Props {
  address: string
  selected: boolean
  balance?: number | null
  onSelect: (address: string) => void
  name?: ElementName
  testID?: string
  hideSelectionCircle?: boolean
}

export const ADDRESS_WRAPPER_HEIGHT = 36
const UNICON_SIZE = 32

export default function WalletPreviewCard({
  address,
  selected,
  balance,
  onSelect,
  hideSelectionCircle,
  ...rest
}: Props): JSX.Element {
  const { name: ensName } = useENS(ChainId.Mainnet, address)
  const isDarkMode = useIsDarkMode()

  const unselectedBorderColor = isDarkMode ? 'none' : 'backgroundOutline'

  return (
    <TouchableArea
      backgroundColor="background1"
      borderColor={selected ? 'magentaVibrant' : unselectedBorderColor}
      borderRadius="rounded20"
      borderWidth={1}
      px="spacing16"
      py="spacing16"
      onPress={(): void => onSelect(address)}
      {...rest}>
      <Flex row alignItems="center" justifyContent="space-between">
        <Flex
          row
          alignItems="center"
          gap="spacing12"
          height={ADDRESS_WRAPPER_HEIGHT}
          justifyContent="flex-start">
          <Unicon address={address} size={UNICON_SIZE} />
          <Box>
            <Text variant="bodyLarge">{ensName ?? shortenAddress(address)}</Text>
            {balance ? (
              <Text color="textSecondary" variant="subheadSmall">
                {formatUSDPrice(balance, NumberType.FiatTokenQuantity)}
              </Text>
            ) : null}
          </Box>
        </Flex>
        {!hideSelectionCircle && (
          <SelectionCircle selected={selected} size="icon16" unselectedColor="textSecondary" />
        )}
      </Flex>
    </TouchableArea>
  )
}
