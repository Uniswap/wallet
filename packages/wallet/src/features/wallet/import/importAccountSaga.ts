import dayjs from 'dayjs'
import { all, call, put } from 'typed-redux-saga'
import { logger } from 'wallet/src/features/logger/logger'
import { Account, AccountType, BackupType } from 'wallet/src/features/wallet/accounts/types'
import { Keyring } from 'wallet/src/features/wallet/Keyring/Keyring'
import { addAccounts, setAccountAsActive, unlockWallet } from 'wallet/src/features/wallet/slice'
import { getValidAddress } from 'wallet/src/utils/addresses'
import { createMonitoredSaga } from 'wallet/src/utils/saga'
import { ImportAccountParams, ImportAccountType } from './types'

export function* importAccount(params: ImportAccountParams) {
  const { type, name } = params
  logger.debug('importAccountSaga', 'importAccount', 'Importing type:', type)

  if (type === ImportAccountType.Address) {
    yield* call(importAddressAccount, params.address, name, params.ignoreActivate)
  } else if (type === ImportAccountType.Mnemonic) {
    yield* call(
      importMnemonicAccounts,
      params.validatedMnemonic,
      params.validatedPassword,
      name,
      params.indexes,
      params.markAsActive,
      params.ignoreActivate
    )
  } else if (type === ImportAccountType.RestoreBackup) {
    yield* call(importRestoreBackupAccounts, params.mnemonicId, params.indexes)
  } else {
    throw new Error('Unsupported import account type')
  }
}

function* importAddressAccount(address: string, name?: string, ignoreActivate?: boolean) {
  const formattedAddress = getValidAddress(address, true)
  if (!formattedAddress) {
    throw new Error('Cannot import invalid view-only address')
  }

  const account: Account = {
    type: AccountType.Readonly,
    address: formattedAddress,
    name,
    pending: true,
    timeImportedMs: dayjs().valueOf(),
  }
  yield* call(onAccountImport, account, ignoreActivate)
}

/**
 * Imports an account with a seed phrase, and handles secure storage.
 * Storage is encrypted via the provided password. Password is discarded
 * after function execution.
 *
 *       [seed]
 *       [ pw ] ┌───────┐    ┌──────────┬────────────────────────┐
 *  User───────►│ Popup │    │Background│                        │
 *              └───┬───┘    ├──────────┘                        │
 *                  │        │  ┌─────────────────┐              │
 *                  └────────┼─►│ImportAccountSaga│              │
 *                    [seed] │  └┬────────────────┘              │
 *                    [ pw ] │   │                               │
 *                           │  ┌▼────────┐ [enc seed]┌─────────┐│
 *                           │  │ Keyring ├──────────►│ Storage ││
 *                           │  └─────┬─▲─┘           └─────────┘│
 *                           │ [seed] │ │ [encrypted seed]       │
 *                           │ [ pw ] │ │                        │
 *                           │      ┌─▼─┴──┐                     │
 *                           │      │Crypto│                     │
 *                           │      └──────┘                     │
 *                           └───────────────────────────────────┘
 */
function* importMnemonicAccounts(
  validatedMnemonic: string,
  validatedPassword?: string,
  name?: string,
  indexes = [0],
  markAsActive?: boolean,
  ignoreActivate?: boolean
) {
  const mnemonicId = yield* call(
    [Keyring, Keyring.importMnemonic],
    validatedMnemonic,
    validatedPassword
  )
  // generate private keys and return addresses for all derivation indexes
  const addresses = yield* all(
    indexes.map((index) => {
      return call([Keyring, Keyring.generateAndStorePrivateKey], mnemonicId, index)
    })
  )

  if (!addresses[0]) throw new Error('Cannot import account with undefined address')
  if (indexes[0] === undefined)
    throw new Error('Cannot import account with undefined derivation index')

  const accounts = addresses.slice(1, addresses.length).map((address, index) => {
    const account: Account = {
      type: AccountType.SignerMnemonic,
      address,
      name,
      pending: true,
      timeImportedMs: dayjs().valueOf(),
      derivationIndex: index + 1,
      mnemonicId,
    }
    return account
  })
  yield* put(addAccounts(accounts))

  const activeAccount: Account = {
    type: AccountType.SignerMnemonic,
    address: addresses[0],
    name,
    pending: !markAsActive,
    timeImportedMs: dayjs().valueOf(),
    derivationIndex: indexes[0],
    mnemonicId,
  }
  yield* call(onAccountImport, activeAccount, ignoreActivate)
}

function* importRestoreBackupAccounts(mnemonicId: string, indexes = [0]) {
  // generate private keys and return addresses for all derivation indexes
  const addresses = yield* all(
    indexes.map((index) => {
      return call([Keyring, Keyring.generateAndStorePrivateKey], mnemonicId, index)
    })
  )
  const accounts = addresses.map((address, index) => {
    const account: Account = {
      type: AccountType.SignerMnemonic,
      address,
      pending: true,
      timeImportedMs: dayjs().valueOf(),
      derivationIndex: index,
      mnemonicId,
      backups: [BackupType.Cloud],
    }
    return account
  })
  yield* put(addAccounts(accounts))
}

function* onAccountImport(account: Account, ignoreActivate?: boolean) {
  yield* put(addAccounts([account]))
  if (!ignoreActivate) {
    yield* put(setAccountAsActive(account.address))
  }
  yield* put(unlockWallet())
  logger.debug('importAccount', '', `New ${account.type} account imported: ${account.address}`)
}

export const {
  name: importAccountSagaName,
  wrappedSaga: importAccountSaga,
  reducer: importAccountReducer,
  actions: importAccountActions,
} = createMonitoredSaga<ImportAccountParams>(importAccount, 'importAccount', {
  showErrorNotification: false,
})
