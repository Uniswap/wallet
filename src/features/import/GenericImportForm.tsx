import { ResponsiveValue, useResponsiveProp } from '@shopify/restyle'
import React, { useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import {
  LayoutChangeEvent,
  LayoutRectangle,
  StyleSheet,
  TextInput as NativeTextInput,
  View,
} from 'react-native'
import { useAppTheme } from 'src/app/hooks'
import AlertTriangle from 'src/assets/icons/alert-triangle.svg'
import PasteButton from 'src/components/buttons/PasteButton'
import { TextInput } from 'src/components/input/TextInput'
import { Box, Flex } from 'src/components/layout'
import { Trace } from 'src/components/telemetry/Trace'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { SectionName } from '../telemetry/constants'

interface Props {
  value: string | undefined
  errorMessage: string | undefined
  onChange: (text: string | undefined) => void
  placeholderLabel: string | undefined
  onSubmit?: () => void
  showSuccess?: boolean // show success indicator
  inputSuffix?: string //text to auto to end of input string
  liveCheck?: boolean
  autoCorrect?: boolean
  inputAlignment?: ResponsiveValue<'center' | 'flex-start', Theme>
  onBlur?: () => void
  onFocus?: () => void
  beforePasteButtonPress?: () => void
  afterPasteButtonPress?: () => void
  blurOnSubmit?: boolean
}

export function GenericImportForm({
  value,
  onChange,
  errorMessage,
  placeholderLabel,
  onSubmit,
  showSuccess,
  inputSuffix,
  liveCheck,
  autoCorrect,
  onBlur,
  onFocus,
  beforePasteButtonPress,
  afterPasteButtonPress,
  blurOnSubmit,
  inputAlignment = 'center',
}: Props): JSX.Element {
  const { t } = useTranslation()
  const theme = useAppTheme()
  const [focused, setFocused] = useState(false)
  const [layout, setLayout] = useState<LayoutRectangle | null>()
  const textInputRef = useRef<NativeTextInput>(null)

  const INPUT_FONT_SIZE = theme.textVariants.bodyLarge.fontSize
  const INPUT_MAX_FONT_SIZE_MULTIPLIER = theme.textVariants.bodyLarge.maxFontSizeMultiplier

  const minHeight = useResponsiveProp({ xs: 90, sm: 120 })
  const px = useResponsiveProp({ xs: 'spacing24', sm: 'spacing36' })
  const py = useResponsiveProp({ xs: 'spacing8', sm: 'spacing16' })

  const handleBlur = (): void => {
    setFocused(false)
    onBlur?.()
  }

  const handleFocus = (): void => {
    setFocused(true)
    onFocus?.()
    // Need this to allow for focus on click on container.
    textInputRef?.current?.focus()
  }

  const handleSubmit = (): void => {
    onSubmit && onSubmit()
  }

  return (
    <Trace section={SectionName.ImportAccountForm}>
      <Flex gap="spacing16" onTouchEnd={handleFocus}>
        <Flex
          centered
          backgroundColor="background1"
          borderColor={
            showSuccess
              ? 'accentSuccess'
              : errorMessage && (liveCheck || !focused) && value
              ? 'accentCritical'
              : 'background2'
          }
          borderRadius="rounded20"
          borderWidth={1}
          flexShrink={1}
          gap="none"
          minHeight={minHeight}
          px={px}
          py={py}
          width="100%">
          {/* TODO: [MOB-3890] make Box press re-focus TextInput. Fine for now since TexInput has autoFocus */}
          <Box
            alignItems="flex-end"
            flexDirection="row"
            justifyContent={inputAlignment}
            width="100%">
            <TextInput
              ref={textInputRef}
              autoFocus
              autoCapitalize="none"
              autoCorrect={Boolean(autoCorrect)}
              backgroundColor="none"
              blurOnSubmit={blurOnSubmit ?? false}
              fontSize={INPUT_FONT_SIZE}
              justifyContent="center"
              lineHeight={INPUT_FONT_SIZE}
              maxFontSizeMultiplier={INPUT_MAX_FONT_SIZE_MULTIPLIER}
              multiline={true}
              numberOfLines={4}
              px="none"
              py="none"
              returnKeyType="done"
              scrollEnabled={false}
              selectionColor={theme.colors.textPrimary}
              spellCheck={false}
              testID="import_account_form/input"
              textAlign={inputAlignment === 'center' || !value ? 'left' : 'center'}
              value={value}
              width={value ? 'auto' : (layout?.width || 0) + theme.spacing.spacing8}
              onBlur={handleBlur}
              onChangeText={onChange}
              onFocus={handleFocus}
              onSubmitEditing={handleSubmit}
            />
            {inputSuffix && value && !value.includes(inputSuffix) ? (
              <View pointerEvents="none">
                <TextInput
                  backgroundColor="none"
                  color="textSecondary"
                  editable={false}
                  fontSize={INPUT_FONT_SIZE}
                  justifyContent="center"
                  lineHeight={INPUT_FONT_SIZE}
                  maxFontSizeMultiplier={INPUT_MAX_FONT_SIZE_MULTIPLIER}
                  multiline={true}
                  px="none"
                  py="none"
                  scrollEnabled={false}
                  value={inputSuffix}
                />
              </View>
            ) : null}
          </Box>
          {!value && (
            <Flex
              centered
              grow
              row
              gap="spacing8"
              position="absolute"
              pt="spacing4"
              onLayout={(event: LayoutChangeEvent): void => setLayout(event.nativeEvent.layout)}>
              <Text
                adjustsFontSizeToFit
                color="textSecondary"
                maxFontSizeMultiplier={INPUT_MAX_FONT_SIZE_MULTIPLIER}
                numberOfLines={1}
                style={styles.placeholderLabelStyle}
                variant="bodyLarge">
                {t('Type or')}
              </Text>
              <PasteButton
                afterClipboardReceived={afterPasteButtonPress}
                beforePress={beforePasteButtonPress}
                onPress={onChange}
              />
              {placeholderLabel && (
                <Text
                  adjustsFontSizeToFit
                  color="textSecondary"
                  numberOfLines={1}
                  style={styles.placeholderLabelStyle}
                  variant="bodyLarge">
                  {placeholderLabel}
                </Text>
              )}
            </Flex>
          )}
        </Flex>
        <Flex>
          {errorMessage && value && (liveCheck || !focused) && (
            <Flex centered row gap="spacing12">
              <AlertTriangle color={theme.colors.accentCritical} />
              <Text color="accentCritical" variant="bodyLarge">
                {errorMessage}
              </Text>
            </Flex>
          )}
        </Flex>
      </Flex>
    </Trace>
  )
}

const styles = StyleSheet.create({
  placeholderLabelStyle: {
    flexShrink: 1,
  },
})
