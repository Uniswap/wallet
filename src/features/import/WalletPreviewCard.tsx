import React from 'react'
import Check from 'src/assets/icons/check.svg'
import { AddressDisplay } from 'src/components/AddressDisplay'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { ElementName } from 'src/features/telemetry/constants'
import { formatUSDPrice, NumberType } from 'src/utils/format'

interface Props {
  address: string
  selected: boolean
  balance?: number | null
  onSelect: (address: string) => void
  name?: ElementName
  testID?: string
}

export const ADDRESS_WRAPPER_HEIGHT = 36
const UNICON_SIZE = 32
const CHECK_ICON_SIZE = 24

export default function WalletPreviewCard({
  address,
  selected,
  balance,
  onSelect,
  ...rest
}: Props): JSX.Element {
  return (
    <TouchableArea
      backgroundColor={selected ? 'background3' : 'background1'}
      borderColor={selected ? 'accentActive' : 'none'}
      borderRadius="rounded16"
      borderWidth={1}
      px="spacing16"
      py="spacing12"
      onPress={(): void => onSelect(address)}
      {...rest}>
      <Flex row alignItems="center" justifyContent="space-between">
        <Flex
          row
          alignItems="center"
          gap="spacing12"
          height={ADDRESS_WRAPPER_HEIGHT}
          justifyContent="flex-start">
          {selected && (
            <Flex centered height={UNICON_SIZE} width={UNICON_SIZE}>
              <Check height={CHECK_ICON_SIZE} width={CHECK_ICON_SIZE} />
            </Flex>
          )}
          <AddressDisplay
            address={address}
            showAccountIcon={!selected}
            size={UNICON_SIZE}
            textAlign="flex-start"
            variant="bodyLarge"
          />
        </Flex>
        {balance ? (
          <Text color="textSecondary" variant="subheadSmall">
            {formatUSDPrice(balance, NumberType.FiatTokenQuantity)}
          </Text>
        ) : null}
      </Flex>
    </TouchableArea>
  )
}
