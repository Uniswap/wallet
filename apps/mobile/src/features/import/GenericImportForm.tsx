import { ResponsiveValue, useResponsiveProp } from '@shopify/restyle'
import React, { useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import {
  LayoutChangeEvent,
  LayoutRectangle,
  StyleSheet,
  TextInput as NativeTextInput,
} from 'react-native'
import { useAppTheme } from 'src/app/hooks'
import PasteButton from 'src/components/buttons/PasteButton'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import Trace from 'src/components/Trace/Trace'
import InputWithSuffix from 'src/features/import/InputWithSuffix'
import { SectionName } from 'src/features/telemetry/constants'
import AlertTriangle from 'ui/src/assets/icons/alert-triangle.svg'
import { Theme } from 'ui/src/theme/restyle/theme'

interface Props {
  value: string | undefined
  errorMessage: string | undefined
  onChange: (text: string | undefined) => void
  placeholderLabel: string | undefined
  onSubmit?: () => void
  showSuccess?: boolean // show success indicator
  inputSuffix?: string //text to auto to end of input string
  liveCheck?: boolean
  autoCorrect?: boolean
  inputAlignment?: ResponsiveValue<'center' | 'flex-start', Theme>
  onBlur?: () => void
  onFocus?: () => void
  beforePasteButtonPress?: () => void
  afterPasteButtonPress?: () => void
  blurOnSubmit?: boolean
  textAlign?: 'left' | 'right' | 'center'
}

export function GenericImportForm({
  value,
  onChange,
  errorMessage,
  placeholderLabel,
  onSubmit,
  showSuccess,
  inputSuffix,
  liveCheck,
  autoCorrect,
  onBlur,
  onFocus,
  beforePasteButtonPress,
  afterPasteButtonPress,
  blurOnSubmit,
  textAlign,
  inputAlignment = 'center',
}: Props): JSX.Element {
  const { t } = useTranslation()
  const theme = useAppTheme()
  const [focused, setFocused] = useState(false)
  const [layout, setLayout] = useState<LayoutRectangle | null>()
  const textInputRef = useRef<NativeTextInput>(null)

  const INPUT_FONT_SIZE = theme.textVariants.bodyLarge.fontSize
  const INPUT_MAX_FONT_SIZE_MULTIPLIER = theme.textVariants.bodyLarge.maxFontSizeMultiplier

  const minHeight = useResponsiveProp({ xs: 90, sm: 120 })
  const px = useResponsiveProp({ xs: 'spacing24', sm: 'spacing36' })
  const py = useResponsiveProp({ xs: 'spacing8', sm: 'spacing16' })

  const handleBlur = (): void => {
    setFocused(false)
    onBlur?.()
  }

  const handleFocus = (): void => {
    setFocused(true)
    onFocus?.()
    // Need this to allow for focus on click on container.
    textInputRef?.current?.focus()
  }

  const handleSubmit = (): void => {
    onSubmit && onSubmit()
  }

  return (
    <Trace section={SectionName.ImportAccountForm}>
      <Flex gap="spacing16" onTouchEnd={handleFocus}>
        <Flex
          centered
          backgroundColor="background1"
          borderColor={
            showSuccess
              ? 'accentSuccess'
              : errorMessage && (liveCheck || !focused) && value
              ? 'accentCritical'
              : 'background2'
          }
          borderRadius="rounded20"
          borderWidth={1}
          flexShrink={1}
          gap="none"
          minHeight={minHeight}
          px={px}
          py={py}
          width="100%">
          {/* TODO: [MOB-225] make Box press re-focus TextInput. Fine for now since TexInput has autoFocus */}
          <InputWithSuffix
            autoCorrect={Boolean(autoCorrect)}
            blurOnSubmit={blurOnSubmit ?? false}
            inputAlignment={inputAlignment}
            inputFontSize={INPUT_FONT_SIZE}
            inputMaxFontSizeMultiplier={INPUT_MAX_FONT_SIZE_MULTIPLIER}
            inputSuffix={inputSuffix}
            layout={layout}
            textAlign={textAlign}
            textInputRef={textInputRef}
            theme={theme}
            value={value}
            onBlur={handleBlur}
            onChangeText={onChange}
            onFocus={handleFocus}
            onSubmitEditing={handleSubmit}
          />
          {!value && (
            <Flex
              centered
              grow
              row
              gap="spacing8"
              position="absolute"
              pt="spacing4"
              onLayout={(event: LayoutChangeEvent): void => setLayout(event.nativeEvent.layout)}>
              <Text
                adjustsFontSizeToFit
                color="textSecondary"
                maxFontSizeMultiplier={INPUT_MAX_FONT_SIZE_MULTIPLIER}
                numberOfLines={1}
                style={styles.placeholderLabelStyle}
                variant="bodyLarge">
                {t('Type or')}
              </Text>
              <PasteButton
                afterClipboardReceived={afterPasteButtonPress}
                beforePress={beforePasteButtonPress}
                onPress={onChange}
              />
              {placeholderLabel && (
                <Text
                  adjustsFontSizeToFit
                  color="textSecondary"
                  numberOfLines={1}
                  style={styles.placeholderLabelStyle}
                  variant="bodyLarge">
                  {placeholderLabel}
                </Text>
              )}
            </Flex>
          )}
        </Flex>
        <Flex>
          {errorMessage && value && (liveCheck || !focused) && (
            <Flex centered row gap="spacing12">
              <AlertTriangle color={theme.colors.accentCritical} />
              <Text color="accentCritical" variant="bodyLarge">
                {errorMessage}
              </Text>
            </Flex>
          )}
        </Flex>
      </Flex>
    </Trace>
  )
}

const styles = StyleSheet.create({
  placeholderLabelStyle: {
    flexShrink: 1,
  },
})
