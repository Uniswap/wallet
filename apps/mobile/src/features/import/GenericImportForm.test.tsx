import React from 'react'
import { GenericImportForm } from 'src/features/import/GenericImportForm'
import { render, screen } from 'src/test/test-utils'

describe(GenericImportForm, () => {
  it('renders a placeholder when there is no value', async () => {
    const tree = render(
      <GenericImportForm
        errorMessage={undefined}
        placeholderLabel="seed phrase"
        value={undefined}
        onChange={(): void => {
          return
        }}
      />
    )

    expect(await screen.findByText('seed phrase')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('renders a value', async () => {
    render(
      <GenericImportForm
        errorMessage={undefined}
        placeholderLabel="seed phrase"
        value="hello"
        onChange={(): void => {
          return
        }}
      />
    )

    expect(await screen.queryByText('seed phrase')).toBeNull()
    expect(await screen.findByDisplayValue('hello')).toBeDefined()
  })

  it('renders an error message', async () => {
    render(
      <GenericImportForm
        errorMessage="there is an error"
        placeholderLabel="seed phrase"
        value="wrong value"
        onChange={(): void => {
          return
        }}
      />
    )

    expect(await screen.findByText('there is an error')).toBeDefined()
  })
})
