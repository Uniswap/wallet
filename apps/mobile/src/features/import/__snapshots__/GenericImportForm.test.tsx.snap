// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericImportForm renders a placeholder when there is no value 1`] = `
<View
  onTouchEnd={[Function]}
  style={
    Array [
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#F5F6FC",
          "borderRadius": 20,
          "borderWidth": 1,
          "display": "flex",
          "flexDirection": "column",
          "flexShrink": 1,
          "justifyContent": "center",
          "minHeight": 120,
          "paddingHorizontal": 36,
          "paddingVertical": 16,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-end",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoComplete="off"
        autoCorrect={false}
        autoFocus={true}
        blurOnSubmit={false}
        maxFontSizeMultiplier={1.4}
        multiline={true}
        numberOfLines={4}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholderTextColor="#888FAB"
        returnKeyType="done"
        scrollEnabled={false}
        selectionColor="#0E1524"
        spellCheck={false}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderRadius": 12,
              "color": "#0E1524",
              "fontSize": 17,
              "justifyContent": "center",
              "lineHeight": 17,
              "paddingHorizontal": 0,
              "paddingVertical": 0,
              "textAlign": "left",
              "width": 8,
            },
          ]
        }
        testID="import_account_form/input"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexGrow": 0,
            "flexShrink": 0,
            "height": 0,
          },
        ]
      }
    />
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "flexGrow": 1,
            "justifyContent": "center",
            "paddingTop": 4,
            "position": "absolute",
          },
        ]
      }
    >
      <Text
        adjustsFontSizeToFit={true}
        allowFontScaling={true}
        maxFontSizeMultiplier={1.4}
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#5D6785",
              "display": "flex",
              "fontFamily": "Inter-Medium",
              "fontSize": 17,
              "lineHeight": 24,
            },
            Object {
              "flexShrink": 1,
            },
          ]
        }
      >
        Type or
      </Text>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 8,
              "width": 8,
            },
          ]
        }
      />
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 5,
            "left": 5,
            "right": 5,
            "top": 5,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#E8ECFB",
            "borderColor": "transparent",
            "borderRadius": 8,
            "borderWidth": 1,
            "display": "flex",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingHorizontal": 8,
                "paddingVertical": 8,
              },
            ]
          }
        >
          <svg
            color="#0E1524"
            data-file-name="SvgPaste"
            height={16}
            strokeWidth={2}
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 4,
                  "width": 4,
                },
              ]
            }
          />
          <Text
            allowFontScaling={true}
            maxFontSizeMultiplier={1.2}
            style={
              Array [
                Object {
                  "color": "#0E1524",
                  "display": "flex",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 15,
                  "lineHeight": 20,
                  "textAlign": "center",
                },
              ]
            }
          >
            Paste
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 8,
              "width": 8,
            },
          ]
        }
      />
      <Text
        adjustsFontSizeToFit={true}
        allowFontScaling={true}
        maxFontSizeMultiplier={1.4}
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#5D6785",
              "display": "flex",
              "fontFamily": "Inter-Medium",
              "fontSize": 17,
              "lineHeight": 24,
            },
            Object {
              "flexShrink": 1,
            },
          ]
        }
      >
        seed phrase
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 16,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "flex-start",
        },
      ]
    }
  />
</View>
`;
